spec/p2p/p2p.qnt

```diff
@@ -0,0 +1,557 @@
+        requestingNodes' = Set(),
+        config' = Map (
+            {ID: "scully", networkAddress: "127.0.0.1177:10"} -> {
+                peerState: Map(),
+                connections: Set(),
+                AddressBook: Set({ID: "igor", networkAddress: "127.0.0.1:10"}, {ID: "josef", networkAddress: "127.0.0.2:10"}),
+                IsUnconditional: Set({ID: "zarko", networkAddress: "127.0.0.42:10"}),
+                PersistentPeers: Set({ID: "bucky", networkAddress: "127.0.0.9:10"}),
+            }
+        )
+    }
+
+    action initEnvEquAddrBook = all {
+        val add = Set({ID: "lasaro", networkAddress: "127.0.0.7:10"}, {ID: "daniel", networkAddress: "127.0.0.6:10"})
+        all {
+        OperatorAddressBook' = Set({ID: "alice", networkAddress: "127.0.0.111:10"}, {ID: "bob", networkAddress: "127.0.0.112:10"}),
+        Environment' = add,
+        requestingNodes' = Set(),
+        config' = Map (
+            {ID: "scully", networkAddress: "127.0.0.1177:10"} -> {
+                peerState: Map(),
+                connections: Set(),
+                AddressBook: add,
+                IsUnconditional: Set({ID: "zarko", networkAddress: "127.0.0.42:10"}),
+                PersistentPeers: Set({ID: "bucky", networkAddress: "127.0.0.9:10"}),
+            }
+        )
+        }
+    }
+
+
+    run happyPathTest = {
+        val a = {ID: "alice", networkAddress: "127.0.0.111:10"}
+        val node = {ID: "scully", networkAddress: "127.0.0.1177:10"}
+        init
+        .then(operatorDialPeers(node , Set(a)))
+        .then(lowerLevelDialPeerSuccess(node, a))
+        .then(all{
+            assert(
+                and {
+                    config.get(node).connections.contains({address: a, state: "outgoing"}),
+                    config.get(node).peerState.get(a) == "outbound",
+                }
+            ),
+            allUnchanged,
+        })
+    }
+
+    //should fail
+    run mixedDirectionsTest = {
+        val a = {ID: "alice", networkAddress: "127.0.0.111:10"}
+        val node = {ID: "scully", networkAddress: "127.0.0.1177:10"}
+        init
+        .then(actionDialPeer(node,a))
+        .then(actionTransportAccept(node,a))
+        .then(lowerLevelTransportAcceptSuccess(node,a))
+        .then(all{
+            assert(invDirectionsOut),
+            allUnchanged,
+        })
+    }
+
+    action step = any {
+        nondet a = oneOf(modeledNodes) 
+        nondet outgoing = oneOf(config.get(a).AddressBook)
+        nondet incoming = oneOf(Environment)
+        nondet conn = oneOf(config.get(a).connections.union(Set({address: {ID: "default", networkAddress: "0.0.0.0"}, 
+                                                                 state: "outgoing"})))
+        nondet dial = oneOf(OperatorAddressBook.powerset())
+        nondet seedResponse = oneOf(Environment.powerset())
+            any {            
+                actionDialPeer(a, outgoing),
```

Comment:
`actionDialPeer` always succeeds, even if `a` is not dialed so this `any` may always succeed here and the action is in practice just a stutter.

Reply:
Yes. I agree. I didn't put much effort in making `step` precise yet. What I also didn't do, for instance, is to set up `step` so that actually two modelled nodes are "connecting". 

First I really would like to get more clarity on what we would like to investigate. Then I would envisage a workflow where we first write tests, and only then think in more detail about making `step` more precise.